man/pycdlib-extract-files.1
.TH PYCDLIB-EXTRACT-FILES 1 "Sep 2018" "pycdlib-extract-files"

.SH NAME
pycdlib-extract-files - tool to extract files from an ISO using pycdlib

.SH SYNOPSIS
.B pycdlib-extract-files [OPTIONS] <iso-file>

.SH DESCRIPTION
This is a tool to extract files from an existing ISO file.  Using this tool,
the files and directories on an ISO can be extracted to the local filesystem.

.SH OPTIONS
.TP
.BI \-path\-type " [auto,iso9660,rockridge,joliet,udf]"
Specifies the path name convention to use while extracting the files.  If the
ISO to be extracted doesn't contain the path type specified, an error is thrown.
If not specified, defaults to auto, which first tries to extract files via the
UDF path, then the Rock Ridge path, then the Joliet path, and finally falls back
to the ISO9660 path if all else fails.  Once one path type succeeds, all files
will be extracted as that path type, and then pycdlib\-extract\-files will quit.
.TP
.BI \-start\-path " <pathname>"
The ISO path to start extracting from, specified in a Unix like path (something
like "/foo/bar").  If not specified, extraction starts from the root of the ISO
(equivalent to specifying "/").  The starting path must be a directory on the
ISO.  If \-path\-type is not specified, the path is an ISO9660 style path.
If \-path\-type is specified, the start path specified here must be of that
particular type.
.TP
.BI \-extract\-to " <destpath>"
The local directory to extract the data to.  The "destpath" must exist and
must be a directory, or an error will be thrown.  If not specified,
pycdlib\-extract\-files will extract the files to the current directory.

.SH SEE ALSO
pycdlib-explorer(1), pycdlib-genisoimage(1)

.SH AUTHOR
Chris Lalancette <clalancette@gmail.com>


CHANGELOG
Changelog for PyCdlib

1.9.0 (2019-12-20)
------------------
* Much improved support for UDF
* Switch to read-only open by default for the open() call
* Fixes to work on big-endian machines
* APIs added:
  * file_mode
* APIs removed:
  None
* APIs deprecated:
  None

1.8.0 (2019-08-10)
------------------
* Addition of "facades" for using PyCdlib in simpler ways
* Make iso_path optional for UDF when using add_symlink
* Make iso_path optional when using rm_file()
* Make iso_path optional when using add_fp()/add_file()
* Create symlinks appropriately in pycdlib-extract-files
* Fix copy_data on macOS
* Support more of the UDF standard
* Cleanup documentation
* Allow UEFI El Torito identifier
* Add an 'auto' mode for pycdlib-extract-files
* Add an '-extract-to' option for pycdlib-extract-files
* APIs added:
  * get_iso9660_facade
  * get_joliet_facade
  * get_rock_ridge_facade
  * get_udf_facade
  * has_rockridge
  * has_joliet
  * has_udf
* APIs removed:
  None
* APIs deprecated:
  None

1.7.0 (2019-02-27)
------------------
* Better UDF compatibility
* Add a tool to extract files with pycdlib
* Greatly improve performance of pycdlib-genisoimage duplicate detection
* Add a design document to the documentation
* Better support for Rock Ridge SF records
* Add type annotation to the entire codebase
* Fix a bug with UDF directory sorting
* Many more unit tests added
* APIs added:
  * walk()
  * open_file_from_iso()
* APIs removed:
  None
* APIs deprecated:
  None

1.6.0 (2018-07-29)
------------------
* More fixes for better UDF compatibility
* Internal revamp of UDF to move some functionality into the classes
* Share more code internally
* More tests
* Better UDF support in pycdlib-explorer
* More documentation, mostly in pycdlib-explorer
* Python 3.4 compatibility
* APIs added:
  None
* APIs removed:
  None
* APIs deprecated:
  None

1.5.0 (2018-06-23)
------------------
* Many more fixes for UDF compatibility
* Support for arbitrarily many hard-links between El Torito, Joliet, ISO9660, and UDF
* Increase test coverage
* Add support for duplicate file detection in pycdlib-genisoimage
* Add support for Rock Ridge version 1.10
* Fix up unicode handling of filenames
* Minor performance optimizations
* APIs added:
  None
* APIs removed:
  None
* APIs deprecated:
  None

1.4.0 (2018-05-04)
------------------
* Lots more updates to pycdlib-genisoimage
* Support for absolute Rock Ridge symlinks
* Initial UDF support
* Documentation updates
* Performance optimizations
* Increase test coverage
* Deprecate a bunch of methods
* Support ISOs that report incorrect PVD sizes
* Initial UDF support
* APIs added:
  * get_file_from_iso()
  * get_file_from_iso_fp()
  * list_children()
  * get_record()
  * set_relocated_name()
* APIs removed:
  None
* APIs deprecated:
  * get_and_write()
  * get_and_write_fp()
  * add_joliet_directory()
  * rm_joliet_directory()
  * list_dir()
  * get_entry()

1.3.2 (2017-11-20)
------------------
* Switch to python3 for the tools
* Small man page fixes
* APIs added:
  None
* APIs removed:
  None
* APIs deprecated:
  None

1.3.1 (2017-11-20)
------------------
* Small fixes to fix up RPM packaging
* APIs added:
  None
* APIs removed:
  None
* APIs deprecated:
  None

1.3.0 (2017-11-19)
------------------
* Performance optimizations
* Support for Joliet levels 1, 2, and 3
* Very large file support (>4GB)
* Increase test coverage
* Add high-level documentation, hosted at https://clalancette.github.io/pycdlib/
* Make API documentation available
* APIs added:
  * add_joliet_directory()
  * rm_joliet_directory()
* APIs removed:
  None
* APIs deprecated:
  None

1.2.0 (2017-09-07)
------------------
* First version of pycdlib-genisoimage, a drop in replacement for genisoimage
* Fixes to allow pycdlib to deal with a lot more files
* Support for floppy and HD booting in El Torito
* Add in an exception hierarchy: PyCdlibException -> PyCdlibInvalidInput, PyCdlibInvalidISO, PyCdlibInternalError
* "Lazy" metadata updating makes this release much faster than previous ones
* Support for more types of ISOs, including those that don't fully conform to standards
* Increase test coverage
* Add lots more functionality to pycdlib-genisoimage
* Lots of optimizations all over the tree
* APIs added:
  * set_hidden()
  * clear_hidden()
  * force_consistency()
* APIs removed:
  None
* APIs deprecated:
  None

1.1.0 (2017-01-31)
------------------
* Better documentation
* Fix up isohybrid and tests
* Start of isohybrid MAC support
* Fix compatibility with Python3 < 3.5
* Cleanup RPM spec file
* Add man pages for the tools
* Add support for very long Rock Ridge filenames
* Lots of fixes to Rock Ridge symlink handling
* Increase test coverage
* APIs added:
  * None
* APIs removed:
  * add_isohybrid_fp()
* APIs deprecated:
  None

1.0.0 (2016-10-25)
------------------
* First stable release with API guarantees
* Rename library to PyCdlib to avoid name clashes
* Many fixes to El Torito handling
* Performance fixes
* Compatibility fixes for ISO9660
* Add an API to allow modifying files in place
* More testing all over the tree
* Add support for ISO9660 interchange level 4
* Add support for "hard-links" between files
* Add support for Rock Ridge 1.12
* Dual Python2/Python3 support
* APIs added:
  * open_fp()
  * get_and_write_fp()
  * write_fp()
  * add_file()
  * modify_file_in_place()
  * add_hard_link()
  * rm_hard_link()
  * add_isohybrid_fp()
  * full_path_from_dirrecord()
  * duplicate_pvd()
* APIs removed:
  * print_tree()
* APIs deprecated:
  None

0.5.0 (2016-01-27)
------------------
* First release of PyCdlib (still called PyISO at the time)
* Initial support for ISO9660, interchange levels 1, 2, and 3
* Initial support for Joliet
* Initial support for Rock Ridge
* Initial support for El Torito
* Initial support for isohybrid
* Created pyiso-compare tool for comparing two ISOs
* Created pyiso-explorer tool for examining ISOs
* APIs added:
  * new()
  * open()
  * print_tree()
  * get_and_write()

tools/pycdlib-extract-files.py
#!/usr/bin/python3

# Copyright (C) 2018  Chris Lalancette <clalancette@gmail.com>

# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation;
# version 2.1 of the License.

# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.

# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

'''
The main code for the pycdlib-extract-files tool, which can extract all or a
subset of files from an ISO.
'''

from __future__ import print_function

import argparse
import collections
import os
import sys

import pycdlib


def parse_arguments():
    '''
    A function to parse all of the arguments passed to the executable.

    Parameters:
     None.
    Returns:
     An ArgumentParser object with the parsed command-line arguments.
    '''
    parser = argparse.ArgumentParser()
    parser.add_argument('-path-type', help='Which path type to use for extraction', action='store', choices=['iso', 'joliet', 'rockridge', 'udf'], default='iso')
    parser.add_argument('iso', help='ISO to open', action='store')
    return parser.parse_args()


def main():
    '''
    The main function for this executable that does the work of extracting
    files from an ISO given the parameters specified by the user.
    '''
    args = parse_arguments()

    iso = pycdlib.PyCdlib()
    print("Opening %s" % (args.iso))
    iso.open(args.iso)

    if args.path_type == 'rockridge':
        if not iso.rock_ridge:
            print('Can only extract Rock Ridge paths from a Rock Ridge ISO')
            return 1
        entry = iso.get_record(rr_path='/')
    elif args.path_type == 'joliet':
        if iso.joliet_vd is None:
            print('Can only extract Joliet paths from a Joliet ISO')
            return 2
        entry = iso.get_record(joliet_path='/')
    elif args.path_type == 'udf':
        if iso.udf_main_descs.pvd is None:
            print('Can only extract UDF paths from a UDF ISO')
            return 3
        entry = iso.get_record(udf_path='/')
    else:
        entry = iso.get_record(iso_path='/')

    dirs = collections.deque([(entry, '/')])
    while dirs:
        dir_record, ident = dirs.popleft()
        ident_to_here = iso.full_path_from_dirrecord(dir_record,
                                                     rockridge=args.path_type == 'rockridge')
        print(ident_to_here)
        if dir_record.is_dir():
            if ident_to_here != '/':
                os.makedirs(ident_to_here[1:])
            if args.path_type == 'rockridge':
                child_lister = iso.list_children(rr_path=ident_to_here)
            elif args.path_type == 'joliet':
                child_lister = iso.list_children(joliet_path=ident_to_here)
            elif args.path_type == 'udf':
                child_lister = iso.list_children(udf_path=ident_to_here)
            else:
                child_lister = iso.list_children(iso_path=ident_to_here)

            for child in child_lister:
                if child is None or child.is_dot() or child.is_dotdot():
                    continue
                dirs.append((child, ident_to_here))
        else:
            if args.path_type == 'rockridge':
                iso.get_file_from_iso(ident_to_here[1:], rr_path=ident_to_here)
            elif args.path_type == 'joliet':
                iso.get_file_from_iso(ident_to_here[1:], joliet_path=ident_to_here)
            elif args.path_type == 'udf':
                iso.get_file_from_iso(ident_to_here[1:], udf_path=ident_to_here)
            else:
                iso.get_file_from_iso(ident_to_here[1:], iso_path=ident_to_here)

    iso.close()
    return 0


if __name__ == "__main__":
    sys.exit(main())
